### ISSUE TITLE
HttpClient - Create synchronous versions of HTTP calls 

### ISSUE THREAD

**Issue by Pastafarian** [dotnet/corefx#13813]
Currently the HttpClient only offers async HTTP calls GetAsync, PostAsync, PutAsync etc... 

When calling from synchronous code it becomes necessary to break to mix async and synchronous code which breaks with best practice and leaves the developer unsure if a ".Result" may result in deadlock.

I suggest we create synchronous versions for all HTTP calls on the client.

**Comment by karelz** [comment]
cc: @davidsh @CIPop @stephentoub

**Comment by svick** [comment]
`Result` resulting in deadlock can only happen if the `Task` you're waiting on attempts to continue on the current context. When it comes to well-written non-UI code, that should never happen. So I think it's safe to assume calling `Result` on the return value of `HttpClient` methods will not deadlock.

**Comment by CIPop** [comment]
> Currently the HttpClient only offers async HTTP calls GetAsync, PostAsync, PutAsync etc...

This is by design.

> When calling from synchronous code it becomes necessary to break to mix async and synchronous code which breaks with best practice and leaves the developer unsure if a ".Result" may result in deadlock.

@Pastafarian the issue that you're pointing out is TPL in general. @stephentoub may be able to address the issue.

**Comment by Pastafarian** [comment]
Thanks for the feedback.

Could you point me in the direction of the reasoning behind only supplying async methods? 

I'm certainly not an expert with async though when I was looking over various articles concerning this (including some by @stephentoub) they appeared to suggest avoiding calling async from synchronous code.

Without looking through the source code for the HttpClient you would be unaware if it was safe to call 'Result' on. I found no documentation suggesting this was the case either.

**Comment by stephentoub** [comment]
> the issue that you're pointing out is TPL in general

It's not even that, it's any synchronous operation needing to consume an asynchronous one, regardless of how it's implemented.

> they appeared to suggest avoiding calling async from synchronous code.

You want to avoid it if possible, but it's not always possible.

> Without looking through the source code for the HttpClient you would be unaware if it was safe to call 'Result' on.

In general it should be "safe" to block waiting for any Task-returning method from the framework; any place that's not the case is a bug that should be fixed.  I put "safe" in quotes because, as with any blocking, you end up burning a thread waiting for the operation to complete, so it's a potential scalability problem.

Note that even if HttpClient were to add synchronous methods, most likely they'd just themselves block waiting for the underlying asynchronous operations to complete.

**Comment by Pastafarian** [comment]
Thanks for explaining @stephentoub, the information regarding the safety of blocking framework tasks is worth knowing.

**Comment by davidsh** [comment]
Given this:

>In general it should be "safe" to block waiting for any Task-returning method from the framework; any place that's not the case is a bug that should be fixed. I put "safe" in quotes because, as with any blocking, you end up burning a thread waiting for the operation to complete, so it's a potential scalability problem.
Note that even if HttpClient were to add synchronous methods, most likely they'd just themselves block waiting for the underlying asynchronous operations to complete.

We think adding sync APIs to the API surface won't really accomplish complete asynchrony.  For use cases that require "sync", awaiting the Task should work.  If there are bugs in this area, as @stephentoub suggests, please open new issues.

**Comment by godefroi** [comment]
> We think adding sync APIs to the API surface won't really accomplish complete asynchrony. For use cases that require "sync", awaiting the Task should work. If there are bugs in this area, as @stephentoub suggests, please open new issues.

@davidsh Can you clarify how this should be done? Using `await` is not possible in sync code, and we've run into the problem where we had the entire thread pool all waiting on `.Result` calls that would never finish, because the tasks couldn't start, (because the thread pool was exhausted waiting for tasks to start).

**Comment by davidsh** [comment]
Commenting on closed issues should be avoided.  If there is a question/bug, please open a new issue.

I'm not sure of what your question really is.  The entire networking stack for HTTP is async throughout.  So, adding new APIs to HttpClient would require not just new APIs but also re-writing portions of the stack to use synchronous socket APIs as well.  Otherwise it really won't be sync.

**Comment by godefroi** [comment]
@davidsh my question is extremely simple; you suggested that "awaiting the Task should work", but I don't know what that means in the context of calling async methods from sync code. Surely you didn't mean the `await` keyword (this being sync code, after all), so what did you mean?

**Comment by godefroi** [comment]
@davidsh Thanks for your clarification. This technique will open us up to thread pool exhaustion, correct? Especially in the case where `GetAsync()` is called from a thread pool thread, it's possible that no thread pool threads are available to actually execute the task, and therefore, `.Result` will hang indefinitely?

**Comment by davidsh** [comment]
@godefroi Using .Result is not a best-practice pattern.  That is why we recommend developers design solutions using async patterns.  That will make the best use of CPU resources.

**Comment by godefroi** [comment]
@davidsh I 100% agree, but there are cases where going to async is not possible. As I understand it (and you haven't indicated otherwise), calling async methods from sync code is not possible to do safely, barring something exotic like a custom non-threadpool-based task scheduler.
